### PR TITLE
Allow floating chunks to be detected in the inspector panel + general refactoring

### LINKED ISSUE #1710: At Cursor feature doesn't detect any toxins
When the cursor hovers over environmental or fired toxins, "At Cursor:" doesn't show you what it is.

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

For reusability, `PlayerHoverInfo` is now renamed into `PlayerInspectInfo`, it is now responsible for the detection of `IInspectableEntity` objects in the game world. This class can be extended by stage-specific derivations to implement stage-specific functions, as exemplified by `MicrobeInspectInfo` where it handle the detection of compound clouds.

Detection is done by raycasting from the camera to what the mouse is pointing at and then checking whatever it intersects. Hopefully with this, the system can be reused as much as possible in other stages for any purpose.

Complementing this entities detection change is the refactor of the hover panel/looking at panel/inspector panel to allow adding detection categories dynamically and consolidating the Node structure of the label controls for detected entities. Descriptions of detected entities can now be easily added, it can either be the amount indicator of compound clouds (_an abundance, a fair amount, etc.._), the n times indicator (_x5, x10, ..._) or any other useful texts to explain to the player what he/she is looking at and what he/she can do with it.

**Related Issues**

Closes #1710.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/general/StageHUDBase.cs
Comment (hhyyrylainen) [inline src/general/StageHUDBase.cs:505]: This should pass `hoverInfoTimeElapsed` as this is time limited so the delta is not really the delta of when this method was last called.

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:14]: Typo

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:20]:
```suggestion
    ///   All inspectable entities the player is pointing at.
```

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:22]: I think this should return an enumerable and leave it up to the caller to decide when they have to make a concrete list out of it.

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:29]: Viewport can be extracteed as a variable.

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:38]: Using the remove all filtering on the previousHits is better I think rather than having to copy the data here.

File: src/general/PlayerInspectInfo.cs
Comment (hhyyrylainen) [inline src/general/PlayerInspectInfo.cs:64]:
```suggestion
    ///   Returns the raycast data of the given raycast inspectable entity.
```
As raycast is the  past tense for raycast.

Comment (hhyyrylainen) [reply src/general/PlayerInspectInfo.cs:64]: Also should this not try to return null when the data doesn't exists, or does it make sense to return blanked out data for non-existent hit?

File: src/gui_common/MouseHoverPanel.cs
Comment (hhyyrylainen) [inline src/gui_common/MouseHoverPanel.cs:55]: Should this be an error level print?

File: src/gui_common/MouseHoverPanel.cs
Comment (hhyyrylainen) [inline src/gui_common/MouseHoverPanel.cs:38]: This could be refactored to be `firstVisibleCategory != null` to avoid having to modify this variable twice (the second modification being after the loop).

File: src/gui_common/MouseHoverPanel.cs
Comment (hhyyrylainen) [inline src/gui_common/MouseHoverPanel.cs:21]:
```suggestion
    private Array categoryControls = new();
```
Would that be a more clear name? From the variable usage I got the impression that this is used for that. The current name `categoriesOrdered` sounds grammatically incorrect so it needs a new name.

File: src/microbe_stage/MicrobeHUD.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:83]: Shouldn't "chunks" be plural here. Also as these are used as keys here, I think either using an enum or a private const variable to hold the string would be better.

With the enum approach the enum needs values for all stages at once in it, but that shouldn't be much of a problem.

File: src/microbe_stage/MicrobeHUD.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:409]: The tuple on these two lines could be a separate variable to not create the tuple twice.

File: src/microbe_stage/MicrobeHUD.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:414]: Why is item nullable? I think it'd be better to force the AddItem to succeed or throw an exception. I noticed this same thing in a few other places in this PR as well. Core code usage problems should throw exceptions to make the methods nicer to use by for example guaranteeing a return value to be non-null.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Seems like a good change. I had some code clarity and style comments (I submitted a few before starting this review). I didn't test this yet but I didn't see any change to AgentProjectile so that leads me to wonder if those are not detected?

File: src/general/PlayerInspectInfo.cs
Comment (Kasterisk) [reply src/general/PlayerInspectInfo.cs:64]: It returns a default.

Comment (hhyyrylainen) [reply src/general/PlayerInspectInfo.cs:64]: Which is a default constructed struct class instance. Which is probably very unhelpful in many cases as knowing between 0 and unset values might be pretty good for not accidentally assuming the both cases (for example sub shape hitting which we already have many bugs about) are the same.

File: src/microbe_stage/MicrobeHUD.cs
Comment (Kasterisk) [reply src/microbe_stage/MicrobeHUD.cs]: This is now fixed with const, enum might feel too restrictive.

**Comment by Kasterisk** [comment]
I didn't make that inspectable because it would be very hard to track the projectiles with your mouse anyway, but no harm in making it so.

File: src/microbe_stage/AgentProjectile.cs
Comment (hhyyrylainen) [inline src/microbe_stage/AgentProjectile.cs:25]: This should probably have json ignore attribute on it. I know the other properties don't really have it that well in this class, but this is saved like other entities...

File: src/microbe_stage/MicrobeHUD.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:92]: This doesn't use the const string?

**Review by hhyyrylainen — APPROVED** [review]
The code is good enough now. I'll merge this tomorrow even if the unnecessary class names in exceptions are not removed tomorrow. I'll give that much time to see if anyone has time to playtest this.

**Comment by hhyyrylainen** [comment]
After merging to my branch I noticed there's a bunch of typo warnings, and some actual typos (this is what I get for not opening this PR in Rider), I'll fix those in my prototype branch which I hope to merge this week.

**Pull request merged**
